regex2nfa.erl

```diff
@@ -0,0 +1,111 @@
+      re2post(string:slice(Input, string:length(InsideParentheses) + 2), lists:merge(Output, PostfixOfInside));
+    CharLiteral -> re2post(without_first(Input), lists:append(Output, [CharLiteral]))
+end.
+
+
+extract_from_parentheses(String) -> [Inside|_] = string:split(string:slice(String, 1), ")", trailing), Inside.
+
+
+post2graph(Input) -> post2graph(Input, initialize_graph(), [], 0).
+post2graph([], Graph, _, _) -> Graph;
+post2graph(Postfix, Graph, Stack, CurrentState) ->
+  case string:slice(Postfix, 0, 1) of
+    "." -> {GraphNew, StackNew, CurrentStateNew} = process_concatenation(Graph, Stack, CurrentState),
+      post2graph(without_first(Postfix), GraphNew, StackNew, CurrentStateNew);
+    "+" -> {GraphNew, StackNew, CurrentStateNew} = process_alternation(Graph, Stack, CurrentState),
+      post2graph(without_first(Postfix), GraphNew, StackNew, CurrentStateNew);
+    "*" -> {GraphNew, StackNew, CurrentStateNew} = process_multiplication(Graph, Stack, CurrentState),
+      post2graph(without_first(Postfix), GraphNew, StackNew, CurrentStateNew);
+    CharLiteral -> post2graph(without_first(Postfix), Graph, [[CharLiteral]|Stack], CurrentState)
+  end.
+
+
+%% Bzdeco we're running form State = 0, and build graph from there
+initialize_graph() ->
+  Graph = digraph:new(),
+  digraph:add_vertex(Graph, {vertex, 0}),
+  Graph.
+
+
+process_concatenation(Graph, [First|Tail], State) ->
+  [Second|Rest] = Tail,
+  CurrentVertex = digraph:vertex(Graph, {state, State}),
+  FirstVertex = digraph:add_vertex(Graph, {vertex, State + 1}),
+  SecondVertex = digraph:add_vertex(Graph, {vertex, State + 2}),
+  digraph:add_edge(Graph, CurrentVertex, FirstVertex, First),
+  digraph:add_edge(Graph, FirstVertex, SecondVertex, Second),
+  {Graph, Rest, State+2}.
+
+
+process_alternation(Graph, [First|Tail], State) ->
+  [Second|Rest] = Tail,
+  CurrentVertex = digraph:vertex(Graph, {state, State}),
+  FirstVertex = digraph:add_vertex(Graph, {vertex, State + 1}),
+  SecondVertex = digraph:add_vertex(Graph, {vertex, State + 2}),
+  ResultVertex = digraph:add_vertex(Graph, {vertex, State + 3}),
+  digraph:add_edge(Graph, CurrentVertex, FirstVertex, First),
+  digraph:add_edge(Graph, CurrentVertex, SecondVertex, Second),
+  digraph:add_edge(Graph, FirstVertex, ResultVertex, eps),
+  digraph:add_edge(Graph, SecondVertex, ResultVertex, eps),
+  {Graph, Rest, State+3}.
+
+
+process_multiplication(Graph, [First|Rest], State) ->
+  CurrentVertex = digraph:vertex(Graph, {state, State}),
+  NewVertex = digraph:add_vertex(Graph, {vertex, State + 1}),
+  AfterMultiplicationVertex = digraph:add_vertex(Graph, {vertex, State + 2}),
+  digraph:add_edge(Graph, CurrentVertex, NewVertex, First),
+  digraph:add_edge(Graph, NewVertex, CurrentVertex, eps),
+  digraph:add_edge(Graph, CurrentVertex, AfterMultiplicationVertex, eps),
+  {Graph, Rest, State+2}.
+
+
+without_first(Input) -> string:slice(Input, 1, string:length(Input)).
```

Comment:
You can just use prolog-style pattern matching for that kind of work - `[First | Rest]` should do.